File: InterviewPreparationKit/CountingValleys.js
// https://www.hackerrank.com/challenges/counting-valleys/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}

/*
 * Complete the 'countingValleys' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER steps
 *  2. STRING path
 */

function countingValleys(steps, path) {
    let altitude = 0;
    let valleys = 0;
    let inValley = false;

    for (let i = 0; i < steps; i++) {
        if (path[i] == 'U') {
            altitude++;
        }
        else {
            altitude--;
        }

        if (altitude < 0) {
            inValley = true;
        }
        else if (inValley) {
            valleys++;
            inValley = false;
        }
    }

    return valleys;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const steps = parseInt(readLine().trim(), 10);

    const path = readLine();

    const result = countingValleys(steps, path);

    ws.write(result + '\n');

    ws.end();
}

File: Interview Preparation Kit/Warm-up Challenges/CountingValleys.py
#!/bin/python3
# https://www.hackerrank.com/challenges/counting-valleys/problem
import math
import os
import random
import re
import sys

#
# Complete the 'countingValleys' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER steps
#  2. STRING path
#

def countingValleys(steps, path):
    altitude = 0
    valleys = 0
    inValley = False

    for i in range(steps):
        if path[i] == 'U':
            altitude += 1
        else:
            altitude -= 1

        if altitude < 0:
            inValley = True
        elif inValley:
            valleys += 1
            inValley = False

    return valleys

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    steps = int(input().strip())

    path = input()

    result = countingValleys(steps, path)

    fptr.write(str(result) + '\n')

    fptr.close()
